RustLeanModels/RustString.lean

```diff
@@ -0,0 +1,4914 @@
+-- Copyright Kani Contributors
+-- SPDX-License-Identifier: Apache-2.0 OR MIT
+import RustLeanModels.Basic
+import RustLeanModels.Iterator
+import Lean
+open List
+open Mathlib
+open Nat
+open Char
+open Iterator
+set_option maxHeartbeats 10000000
+
+/-
+Variable naming conventions:
+-Assumptions (Prop): starts with g  (I save h for h::t)
+-Strings (Str): s, s1, s2, (h::t)
+-Substrings (Str): ss
+-Prefix (Str): p p1 p2
+-Character: c
+-Pattern: P
+-Char filter (Char → Bool): f
+-Position, Index (Nat) : i j
+-Word (member of a split) (Str) : w
+-General Nat: m n
+-List nat: l
+-Element of list, input of function: x
+-Extra parameter in aux functions: k
+-/
+
+namespace RustString
+
+
+def rust_isWhitespace (c : Char) := (isWhitespace c) || (c = '\x0C')
+
+
+def is_Unicode_Whitespace (c : Char): Bool :=
+  c.val ∈ [9, 10, 11, 12, 13, 32, 133, 160, 5760,
+          8192, 8193, 8194, 8195, 8196, 8197, 8198, 8199, 8200, 8201, 8202,
+          8232, 8233, 8239, 8287, 12288]
+
+def Char_is_ascii (c : Char) : Bool := c.val ≤ 127
+
+/-
+All iterator's functions are implemented directly on Str
+
+chars => s.chars() = s
+
+-/
+
+/-
+Function: byteSize
+From: core::str::len
+Description: https://doc.rust-lang.org/std/primitive.str.html#method.len
+-/
+def sum_list_Nat (l : List Nat) := List.foldl Nat.add 0 l
+
+lemma fold_add_para1_elim: List.foldl Nat.add x l = x + List.foldl Nat.add 0 l := by
+  induction l generalizing x
+  simp only [foldl_nil, add_zero]
+  rename_i h t ind
+  simp only [foldl_cons, add_eq, zero_add] ; rw[@ind h, @ind (x+h)]; omega
+
+lemma sum_List_Nat_cons : sum_list_Nat (h::t) = h + sum_list_Nat t  :=by
+  simp only [sum_list_Nat, foldl_cons, add_eq, zero_add]
+  exact fold_add_para1_elim
+
+
+@[simp]
+def byteSize (s : List Char) : Nat := match s with
```

Comment:
It's useful to add a description of those functions in the form of doc comments, e.g.
```lean
/--
Compute the length of `s` in terms of the number of bytes.
The length is the sum of the number of bytes for each individual character.
-/